Exceptions thrown by extensions invisible in Maven build logs

### Describe the bug

As mentioned first in https://github.com/spockframework/spock/pull/1437#issuecomment-1066028196 and then in https://github.com/spockframework/spock/pull/1442#discussion_r825391679, if an annotation-driven extension throws a runtime exception, e.g. `InvalidSpecException` or `ExtensionException`, this is not reflected in the Maven build log, not even in debug mode. So I guess that either the JUnit 5 platform (my guess) or some part in Spock is swallowing those exceptions.

When running the same spec in an IDE like IntelliJ IDEA, the extension exception is shown. In Maven ~~and probably also in Gradle~~, the whole affected spec is simply not added to the test plan, i.e. it is not reported as ignored or failed at all. Update: In Gradle, the spec is reported as failed.

This could affect a large number of built-in Spock extensions throwing exceptions, see the attached [search log](https://github.com/spockframework/spock/files/8239056/report.txt) from my IDE.


### To Reproduce

Add Spock annotations provoking errors to a test spec, e.g. a combination of `@Unroll` and `@Rollup` on the same spec or feature method. Run it, first from an IDE, then from Maven. _(Continued in "actual behaviour" section.)_

### Expected behavior

* The exception thrown by the extension is reported in the build log.
* The spec or feature is question is reported as a test error (not a failure, because no test was executed).
* The remaining specs or features are executed normally.

### Actual behavior

Run the spec from IntelliJ IDEA. You should see something like
```text
org.spockframework.runtime.InvalidSpecException: @Unroll and @Rollup must not be used on the same feature: run all tests with skip after first failure

  at org.spockframework.runtime.extension.builtin.UnrollExtension.visitFeature(UnrollExtension.java:74)
  at org.spockframework.runtime.extension.builtin.UnrollExtension.lambda$visitSpec$0(UnrollExtension.java:59)
```

Now run the same test from Maven, using a command like `mvn test -DfailIfNoTests=false -Dtest=StepwiseIterationsTest`. You should see something like:
```text
[INFO] --- maven-surefire-plugin:2.22.2:test (default-test) @ perform-tests ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

### Java version

JDK 17, but I do not think it matters.

### Buildtool version

The problem occurs in Maven, not in Gradle.

* Maven 3.8.1
* Surefire 2.22.2 or latest Surefire 3.0.0-M6-SNAPSHOT, same behaviour
* It is independent of whether the compilation is done by GMavenPlus Plugin or by Maven Compiler Plugin with Groovy-Eclipse. In my own projects I always use the latter, while the official Spock example project uses the former. The effect is the same.

### What operating system are you using

Windows

### Dependencies

x

### Additional context

x

If we look at [`org.apache.maven.surefire.junitplatform.TestPlanScannerFilter#accept`](https://github.com/apache/maven-surefire/blob/68bca29b23d27dba3e06cf5e73a3fdc1a22598db/surefire-providers/surefire-junit-platform/src/main/java/org/apache/maven/surefire/junitplatform/TestPlanScannerFilter.java#L48-L59), we see:

```java
    @Override
    @SuppressWarnings( "rawtypes" )
    public boolean accept( Class testClass )
    {
        LauncherDiscoveryRequest discoveryRequest = request()
                        .selectors( selectClass( testClass.getName() ) )
                        .filters( includeAndExcludeFilters ).build();
        TestPlan testPlan = launcher.discover( discoveryRequest );
        return testPlan.containsTests();
    }
```

Somehow, for an `ErrorSpecNode` the condition `testPlan.containsTests() is `false`. I experimentally added

```java
  @Override
  public boolean isTest() {
    return true;
  }
```

to `ErrorSpecNode`, and run the example project against a Spock 2.2 snapshot, I see
```text
[INFO] Running DataDrivenSpec
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.002 s <<< FAILURE! - in DataDrivenSpec
[ERROR] DataDrivenSpec  Time elapsed: 0.002 s  <<< ERROR!
org.spockframework.runtime.InvalidSpecException: @Unroll and @Rollup must not be used on the same spec: DataDrivenSpec
(...)
[ERROR] Errors:
[ERROR]   DataDrivenSpec » InvalidSpec @Unroll and @Rollup must not be used on the same ...
[INFO]
[ERROR] Tests run: 35, Failures: 0, Errors: 1, Skipped: 0
```

That looks better, and in Gradle the output stays the same as before, so superficially nothing seems to be broken. But I am not sure if this is the right change. It was just a hack, and I do not know enough about the super-classes and -interfaces implemented by `ErrorSpecNode` to actually judge if this is an correct and sufficient change. I do hope that one of the three people I notified by mentioning their names before can provide additional insights here.

--- a/spock-core/src/main/java/org/spockframework/runtime/ErrorSpecNode.java
+++ b/spock-core/src/main/java/org/spockframework/runtime/ErrorSpecNode.java
@@ -1,5 +1,7 @@
 package org.spockframework.runtime;
 
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.launcher.TestPlan;
 import org.spockframework.runtime.model.SpecInfo;
 import org.spockframework.util.ExceptionUtil;
 import spock.config.RunnerConfiguration;
@@ -14,30 +16,43 @@ public class ErrorSpecNode extends SpecNode {
     this.error = error;
   }
 
+  /**
+   * Prevent this node from being pruned. The default logic would prune it, as it
+   * <ul>
+   *   <li>is no test,</li>
+   *   <li>has no test descendents and</li>
+   *   <li>may not register new tests during execution</li>
+   * </ul>
+   *
+   * Without this empty override, the node is thrown away and the error is not reported.
+   */
   @Override
-  public void prune() {
-    // prevent pruning of this node
-    // default logic would prune it as it
-    // - is no test,
-    // - has no test descendents and
-    // - may not register new tests during execution
-    // without this empty override, the node is thrown away and the error is not reported
-  }
+  public void prune() {}
 
+  /**
+   *  Prevent this node from being removed. As it does not have any children, it would get removed when trying to select
+   *  specific tests methods. For example, Gradle will report that no test were found, and not report the actual error.
+   */
   @Override
-  public void removeFromHierarchy() {
-    // prevent removal of this node
-    // As the ErrorSpecNode does not have children it would get removed when trying to select specific tests methods.
-    // For example, gradle will report that no test were found, and not report the actual error.
-  }
+  public void removeFromHierarchy() {}
 
+  /**
+   * Maven Surefire removes nodes if {@link TestPlan#containsTests()} is <tt>false</tt>.
+   * In order to avoid this, we return <tt>true</tt> here, because it makes {@link #containsTests(TestDescriptor)}
+   * also return <tt>true</tt>, which in turn feeds into Maven's test plan.
+   * This seems to work and not to bother Gradle either.
+   *
+   * @return always <tt>true</tt> for error spec nodes
+   */
   @Override
   public boolean mayRegisterTests() {
-    // Maven Surefire removes nodes if org.junit.platform.launcher.TestPlan.containsTests is false.
-    // The easiest way to avoid this, is to return true here. This seems to work and not to bother Gradle either.
     return true;
   }
 
+  /**
+   * Always throws the error specified in {@link #ErrorSpecNode(UniqueId, RunnerConfiguration, SpecInfo, Throwable)},
+   * escalating it up the call stack.
+   */
   @Override
   public SpockExecutionContext prepare(SpockExecutionContext context) throws Exception {
     return ExceptionUtil.sneakyThrow(error);

--- a/spock-core/src/main/java/org/spockframework/runtime/ErrorSpecNode.java
+++ b/spock-core/src/main/java/org/spockframework/runtime/ErrorSpecNode.java
@@ -32,6 +32,13 @@ public class ErrorSpecNode extends SpecNode {
   }
 
   @Override
+  public boolean mayRegisterTests() {
+    // Maven Surefire removes nodes if org.junit.platform.launcher.TestPlan.containsTests is false.
+    // The easiest way to avoid this, is to return true here. This seems to work and not to bother Gradle either.
+    return true;
+  }
+
+  @Override
   public SpockExecutionContext prepare(SpockExecutionContext context) throws Exception {
     return ExceptionUtil.sneakyThrow(error);
   }

--- a/spock-specs/src/test/groovy/org/spockframework/runtime/ErrorSpecNodeSpec.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/runtime/ErrorSpecNodeSpec.groovy
@@ -33,4 +33,16 @@ class ErrorSpecNodeSpec extends Specification {
     then:
     0 * tdParent.removeChild(*_)
   }
+
+  def 'make TestDescriptor think that an ErrorSpecNode contains tests'() {
+    given:
+    def errorSpecNodeSpy = Spy(errorSpecNode)
+
+    when:
+    def containsTests = TestDescriptor.containsTests(errorSpecNodeSpy)
+
+    then:
+    containsTests
+    1 * errorSpecNodeSpy.mayRegisterTests()
+  }
 }

--- a/spock-specs/src/test/groovy/org/spockframework/runtime/ErrorSpecNodeSpec.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/runtime/ErrorSpecNodeSpec.groovy
@@ -1,6 +1,6 @@
 package org.spockframework.runtime
 
-import spock.lang.Specification
+import spock.lang.*
 
 import org.junit.platform.engine.*
 import spock.lang.Subject
@@ -34,6 +34,7 @@ class ErrorSpecNodeSpec extends Specification {
     0 * tdParent.removeChild(*_)
   }
 
+  @Issue("https://github.com/spockframework/spock/issues/1444")
   def 'make TestDescriptor think that an ErrorSpecNode contains tests'() {
     given:
     def errorSpecNodeSpy = Spy(errorSpecNode)

